### DISCUSSION TITLE
Correct handling of `/` in `pattern`

### DISCUSSION THREAD

**Discussion by tschmidtb51** [json-schema-org/community#684] [Q&A]
I'm wondering whether the following pattern `^ANY:/[ab]$` is a valid JSON pattern. Many implementations accept it but for example Deno requires to escape the `/` for it to be valid. So the question is: Is that an issue with the pattern or with deno? How should the spec be interpreted? (I had a look through the spec but was not really successful to resolve the question).

**Comment by gregsdennis** [comment]
By "Deno" I imagine you mean https://deno.com/?  This seems more like a question about how Javascript handles regex rather than a JSON Schema question.

Can you elaborate?

**Reply by tschmidtb51** [reply]
> By "Deno" I imagine you mean https://deno.com/? 

Yes. Sorry - I should have linked that.

**Reply by tschmidtb51** [reply]
> This seems more like a question about how Javascript handles regex rather than a JSON Schema question.
>
> Can you elaborate?

Sure. IMHO, this is a JSON question:
I created a [GitHub repo](https://github.com/tschmidtb51/JSON-schema-pattern) to back the question with code...

I my reading of the JSON spec, `/` is not a special character in a pattern and therefore should not be needed to be escaped. For example, the reference implementation of the [CSAF](https://csaf.io) [validation library](https://github.com/secvisogram/csaf-validator-lib) uses the [JSON schema provided by the CSAF standard](https://docs.oasis-open.org/csaf/csaf/v2.0/csaf_json_schema.json). There, the `/` was not escaped and it worked with python and node. However, when a colleague used deno, he complained about the "missing" escaping (and another bug; see https://github.com/secvisogram/csaf-validator-lib/issues/156 for the full story).

So my question is: Do I need to escape `/` in JSON patterns or is that a bug in the deno implementation when they convert the JSON pattern into their regex?

Does that make sense?

**Reply by tschmidtb51** [reply]
For clarification: 
CSAF: [Common Security Advisory Framework](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html) - an international, open standard for machine-readable security information (including, but not limited to security advisories, VEX and informational advisories).

**Reply by gregsdennis** [reply]
We don't control anything about JSON here.  We're only JSON Schema, which operates on the JSON _data model_ not on the encoding itself.

That said, it sounds like a bug in Deno to me.  `/` isn't _required_ to be escaped by JSON.  It exists in pointers all of the time.  I'm not sure about whether a Javascript regex would require it to be escaped, though.

**Reply by jdesrosiers** [reply]
JSON doesn't have regular expressions, so it has nothing to say about the matter.

JSON Schema has regular expressions expressed as strings. JSON Schema implementations are expected to use whatever regex engine is available in the language the implementation is written in. Since you mention deno, that means this is a JavaScript question. Specifically, the question is, does `new Regex()` in JavaScript require `/` to be escaped? I tried it in node.js, deno, and firefox and in all cases, the answer is "no", it does not need to be escaped.

**Reply by tschmidtb51** [reply]
Thank you for your answer. We will consider your answer. For the future, it would be nice, if that could be clarified in the standard and would not depend upon the implementation. That would make it easier to write a schema that is interoperable with "all" implementations.

**Reply by gregsdennis** [reply]
[Core 6.4](https://json-schema.org/draft/2020-12/draft-bhutton-json-schema-01#name-regular-expressions) does address the disparity and contains text to encourage compatible expressions. 

> Furthermore, given the high disparity in regular expression constructs support, schema authors SHOULD limit themselves to the following regular expression tokens:
>
> - individual Unicode characters, as defined by the JSON specification [RFC8259];
> - simple character classes ([abc]), range character classes ([a-z]);
> - complemented character classes ([^abc], [^a-z]);
> - simple quantifiers: "+" (one or more), "*" (zero or more), "?" (zero or one), and their lazy versions ("+?", "*?", "??");
> - range quantifiers: "{x}" (exactly x occurrences), "{x,y}" (at least x, at most y, occurrences), {x,} (x occurrences or more), and their lazy versions;
> - the beginning-of-input ("^") and end-of-input ("$") anchors;
> - simple grouping ("(...)") and alternation ("|").

JSON Schema can't _require_ these things for all languages until all languages unify their regex support.
